Add support for Jetty Reactive Streams HTTP client

sdeleuze opened the pull request:
Ready to be reviewed before merging in `master` since I would like to begin having feedback with real applications using Spring Framework snapshots.

Notice that this PR is implemented with buffer copy instead of optimized buffer wrapping because the latter hangs since `Callback#succeeded` doesn't allow releasing the buffer and requesting more data at different times (required for `Mono<DataBuffer>` for example).

Simone is aware of that issue and will see if it can be changed on Jetty side (not sure yet). That could be changed with an additional commit if that becomes possible in the future.

Issue: [SPR-15092](https://jira.spring.io/browse/SPR-15092)

sbordet commented on spring-web/spring-web.gradle:
```diff
@@ -36,6 +36,7 @@ dependencies {
 	optional("io.netty:netty-all")
 	optional("io.projectreactor.ipc:reactor-netty")
 	optional("org.apache.tomcat.embed:tomcat-embed-core:${tomcatVersion}")
+	optional("org.eclipse.jetty:jetty-reactive-httpclient:0.9.2")
```
Version 1.0.0 with RS TCK implemented is available in Maven Central.

sdeleuze replied:
Updated

sdeleuze committed:
Add support for Jetty Reactive Streams HTTP client

Implemented with buffer copy instead of optimized buffer wrapping
because the latter hangs since Callback#succeeded doesn't allow
releasing the buffer and requesting more data at different times
(required for Mono<DataBuffer> for example).
See https://github.com/eclipse/jetty.project/issues/2429.

Issue: SPR-15092

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
+		Publisher<ContentChunk> chunks = Flux.from(publisher).map(DataBuffer::asByteBuffer).map(ContentChunk::new);
```
Even though this class is using `DefaultDataBuffer`, which does not require releasing, I still think it would be a good idea to release them here anyway, especially if we make the data buffer factory configurable.

I think the release can be implemented as the `ContentChunk` callback (I am assuming that the callback would be called after the data has been written, not sure about that). I see that below you said that releasing in the callback does not work for the response, but perhaps it does for the request?

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+		ReactiveRequest.Content content = ReactiveRequest.Content.fromPublisher(Flux
+				.from(body)
+				.flatMap(publisher -> publisher)
+				.map(buffer -> new ContentChunk(buffer.asByteBuffer())), contentType);
```
Release needs to happen here as well.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+		this.response = Mono.from(
+				this.reactiveRequest.response((reactiveResponse, contentChunks) -> {
+
+					// Implementation with buffer copy instead of optimized buffer wrapping
```
Not exactly sure what is meant by this, let's discuss it during our meeting.

sdeleuze replied:
With current Jetty client implementation, we can never transform a `Flux<DataBuffer>` to `Mono<DataBuffer>` because the callback calls `request(1)` + release the buffer. In our case we want to `request(1)` ( to get next buffer) during buffer composition but do the release only when `PooledDataBuffer#release` is called, which is impossible with current API. See the related https://github.com/eclipse/jetty.project/issues/2429 Jetty issue.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+
+	@Override
+	protected void applyHeaders() {
+		getHeaders().entrySet().forEach(e -> e.getValue().forEach(v -> this.jettyRequest.header(e.getKey(), v)));
```
Use a variable instead of calling getHeaders() twice.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+	@Override
+	public HttpMethod getMethod() {
+		HttpMethod method = HttpMethod.resolve(this.jettyRequest.getMethod());
+		Assert.notNull(method, "Method must not be null");
```
Seems like `Assert.state` is more suitable here, because `notNull` results in an `IllegalArgumentException`.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,146 @@
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+public class JettyClientHttpRequest extends AbstractClientHttpRequest {
```
Could be package-protected instead, no need to expose this class.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpResponse.java:
```diff
@@ -0,0 +1,88 @@
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+public class JettyClientHttpResponse implements ClientHttpResponse {
```
Could be package-protected instead, no need to expose it.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpResponse.java:
```diff
@@ -0,0 +1,88 @@
+	private final Flux<DataBuffer> content;
+
+
+	public JettyClientHttpResponse(ReactiveResponse reactiveResponse, Publisher<DataBuffer> content) {
```
The `Publisher<DataBuffer>` parameter should be `Flux<DataBuffer>`, since the only place you're creating them is internally, from `JettyClientHttpRequest`.

poutsma commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpResponse.java:
```diff
@@ -0,0 +1,88 @@
+		Assert.notNull(reactiveResponse, "reactiveResponse should not be null");
+		Assert.notNull(content, "content should not be null");
+		this.reactiveResponse = reactiveResponse;
+		this.content = Flux.from(content);
```
See above, drop the `Flux.from`.

poutsma reviewed (commented):
Looks good overall. Some things that need to change and some things I need more info on during our meeting.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpConnector.java:
```diff
@@ -0,0 +1,122 @@
+	public Mono<ClientHttpResponse> connect(HttpMethod method, URI uri,
+			Function<? super ClientHttpRequest, Mono<Void>> requestCallback) {
+
+		if (!httpClient.isStarted()) {
```
Really minor, but perhaps delegate to `isRunning()`?

sdeleuze replied:
I don't think so, `isStarted()` is different than `isRunning()`.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpConnector.java:
```diff
@@ -0,0 +1,122 @@
+	public int getPhase() {
+		return Integer.MAX_VALUE;
+	}
+}
```
For ordering of `SmartLifecycle` methods, in the very least:
- stop(Runnable) should be after stop()
- getPhase and isAutoStartup up ahead among accessors

As lifecycle methods, personally I like to see start/stop/isRunning ahead of runtime methods like `connect`, and it's something I've been following, but I'm not sure if there a broader precedent.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpConnector.java:
```diff
@@ -0,0 +1,122 @@
+	@Override
+	public void start() {
+		try {
+			this.httpClient.start();
```
A code level comment that explains HttpClient is internally synchronized and protected with state checks around here could be useful.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpConnector.java:
```diff
@@ -0,0 +1,122 @@
+
+		if (!uri.isAbsolute()) {
+			return Mono.error(new IllegalArgumentException("URI is not absolute: " + uri));
+		}
```
No practical difference but maybe this check first (before starting the client above) ?

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+		headers.entrySet().forEach(e -> e.getValue().forEach(v -> this.jettyRequest.header(e.getKey(), v)));
+		if (!headers.containsKey(HttpHeaders.ACCEPT)) {
+			this.jettyRequest.header(HttpHeaders.ACCEPT, "*/*");
+		}
```
Is it necessary to add `"*/*"` explicitly? It should be the same as not having it.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+	public Mono<ClientHttpResponse> getResponse() {
+		return this.response;
+	}
+
```
The overall method ordering should be: accessors, public contract methods, protected contract (base class) methods (e.g. see `ReactorHttpRequest`).

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> publisher) {
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
```
Is it not enough to look up via `getHeaders().getContentType()` ?

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+				DataBufferUtils.release(buffer);
+				throw Exceptions.propagate(x);
+			}
+		}));
```
This could be extracted:
```
private ContentChunk toContentChunk(DataBuffer buffer) {
	return new ContentChunk(buffer.asByteBuffer(), new Callback() {
		@Override
		public void succeeded() {
			DataBufferUtils.release(buffer);
		}

		@Override
		public void failed(Throwable x) {
			DataBufferUtils.release(buffer);
			throw Exceptions.propagate(x);
		}
	});
}
```

Then it becomes:
```
Publisher<ContentChunk> chunks = Flux.from(publisher).map(this::toContentChunk);
```
And the writeAndFlushWith method below does the same AFAICS.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+			}
+		}));
+		ReactiveRequest.Content requestContent = ReactiveRequest.Content.fromPublisher(chunks, contentType);
+		this.reactiveRequest = ReactiveRequest.newBuilder(jettyRequest).content(requestContent).build();
```
Does `reactiveRequest` really need to be a field? It seems like it should be possible to pass it directly into the method called `completes()`. It takes more effort to understand how it's used when it is a field.

rstoyanchev commented on spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java:
```diff
@@ -0,0 +1,174 @@
+
+	public Mono<ClientHttpResponse> getResponse() {
+		return this.response;
+	}
```
It's not obvious at first why a request object should have a method that returns the response, and a public method for that. It must be called at the right time too, or else the response field may not have been initialized, but I can see why the method is here, based on the way the Jetty API works.

I think it would help clarity if the method called `completes` simply stored `ReactiveRequest`, returned `Mono.empty()`, and nothing more. The method already returns `Mono.empty()` anyway, so the part about creating the response is completely independent and could be moved out of the request. In other words, the connector would access the `ReactiveRequest` (via package private method) and then create the response. Then would make it a little more instantly obvious what's going on.

The other point is that the `Mono<Void>` from writeWith is supposed to reflect the completion of the writing of the request body, per contract, but probably not a practical issue at this time. If we keep all this package private we have the option to evolve it in the future.

rstoyanchev commented:
Sorry for the delayed response, but my comments above. Let's get this ready for master, and continue there as needed.

sdeleuze commented:
@rstoyanchev  Thanks for your feedback, I think I have taken your remarks in account. I have not yet merged it in `master` because I have a strange remaining issue: `SseIntegrationTests` and `WebClientIntegrationTests` fail on command line with `./gradlew clean test` (`AssertionError`, likely timeouts) but succeed in IntelliJ IDEA. Any idea?

sdeleuze commented:
Merged via https://github.com/spring-projects/spring-framework/commit/a87764f1fd218356690e05cf2aa964e92202eed6.
